src/components/chip/index.tsx

```diff
@@ -0,0 +1,345 @@
+
+/**
+ * @description: Chip component
+ * @extends: TouchableOpacity
+ * @extendslink: docs/TouchableOpacity
+ * @example: https://github.com/wix/react-native-ui-lib/blob/master/demo/src/screens/componentScreens/ChipScreen.tsx
+ */
+const Chip: React.FC<ChipProps> = ({
+  avatar,
+  backgroundColor,
+  borderRadius = BorderRadiuses.br100,
+  badgeProps,
+  counterLabel,
+  counterColor,
+  counterStyle,
+  containerStyle,
+  dismissColor,
+  dismissIcon,
+  dismissIconStyle,
+  iconColor,
+  iconSource,
+  iconStyle,
+  label,
+  labelStyle,
+  labelColor,
+  minSize,
+  onDismiss,
+  dismissContainerStyle,
+  onPress,
+  resetSpacings,
+  size,
+  testID,
+  useBadge,
+  ...others
+}) => {
+  Chip.displayName = 'Chip';
+  Chip.defaultProps = {
+    borderRadius: BorderRadiuses.br100,
+    counterColor: Colors.grey20,
+    dismissIcon: Assets.icons.x,
+    minSize: true,
+    size: 24
+  };
+
+  const renderIcon = () => {
+    return (
+      <Image
+      // @ts-ignore
+        source={iconSource}
+        tintColor={iconColor}
+        style={[iconStyle, getMargins('iconSource')]}
+        testID={`${testID}.icon`}
+      />
+    );
+  };
+
+  const renderCounter = () => {
+    if (useBadge) {
+      return (
+        <Badge
+          label={counterLabel}
+          backgroundColor={Colors.red30}
+          size={BADGE_SIZES.medium}
+          style={[getMargins('counter')]}
+          labelColor={counterColor}
+          {...badgeProps}
+          {...counterStyle}
+          testID={`${testID}.counter`}
+        />
+      );
+    }
+    return (
+      <Text
+        text90R
+        color={counterColor}
+        style={[getMargins('counter'), counterStyle]}
+        testID={`${testID}.counter`}
+      >
+        {counterLabel}
+      </Text>
+    );
+  };
+
+  const renderOnDismiss = () => {
+    return (
+      <TouchableOpacity
+        style={[getMargins('dismiss'), dismissContainerStyle]}
+        onPress={onDismiss}
+        hitSlop={{top: 10, bottom: 10, left: 10, right: 10}}
+        testID={`${testID}.dismiss`}
+      >
+        <Image
+          source={dismissIcon}
+          tintColor={dismissColor}
+          style={[dismissIconStyle]}
+          accessibilityLabel="dismiss"
+          testID={`${testID}.dismiss`}
+        />
+      </TouchableOpacity>
+    );
+  };
+
+  const renderAvatar = () => {
+    return (
+      <Avatar
+        size={20}
+        {...avatar}
+        // @ts-ignore
+        containerStyle={[getMargins('avatar'), avatar.containerStyle]}
+        testID={`${testID}.avatar`}
+      />
+    );
+  };
+
+  const renderLabel = () => {
+    return (
+      <Text
+        text90M
+        numberOfLines={1}
+        // @ts-ignore
+        tintColor={labelColor}
+        style={[styles.label, getMargins('label'), labelStyle]}
+        testID={`${testID}.label`}
+      >
+        {label}
+      </Text>
+    );
+  };
+
+  const getMargins = (element: string): object | undefined => {
```

Comment:
I'm not a fan of the those huge hub functions, I think we can do a little refactor here: 

First, I see that all elements excepts `label` barely have any logic, so I think we can just set these margins inline in each appropriate render function and consider making this function solely for label. 

Now regarding the label margins logic. 
Since you have a separate render function for each component, wouldn't it be cleaner to check in each function if we have a label and add the margin accordingly to the element (avatar, badge, etc..) and not the label.

Reply:
In general, by refactoring (shortening it) we lose all its purpose.

Which is to avoid specific margin annotations that are disconnected from the rest of the margin logic making it hard for to track, debug and override
Main advantages:
- Readability (all in one place): you know where to look and what effects do the inner elements have on each other in terms of spacings. Instead of hopping from render to render.
- Code cleanliness: no more messy inline styling! (was a fan).
- Supports how the "margins api" is built: we supply basic ui (margins included) and 2 override "switches": you can override specific margin with styling or burn them all with fire and pass `resetMargins`. The hub makes it manageable to see instantly what happens when you have elements together.
Disadvantage: its long.

*I made this choice because to write this logic, I had to decrypt all the margins of the 3 components we want to render in Private using this base component (chip, label, tag). Their margin logic was hell to keep track of.

As I see it, it's either this or going back to inline margins and one function for the label margins `getLabelMargin()`.  Once you add label margins to each component's render function you get 3 and sometimes 4 margin variations (like for counter)